[PATCH] [clang][frontend] Fix serialization for CXXConstructorDecl (refs llvm#132794)

koplas opened the pull request:
When retrieving the ExplicitSpecifier from a CXXConstructorDecl, one of its canonical declarations is returned. To correctly write the declaration record the ExplicitSpecifier of the current declaration must be used.

Failing to do so results in a crash during deserialization.

cor3ntin commented on clang/include/clang/AST/DeclCXX.h:
```diff
@@ -2601,10 +2601,11 @@ class CXXConstructorDecl final
   size_t numTrailingObjects(OverloadToken<InheritedConstructor>) const {
-    return CXXConstructorDeclBits.IsInheritingConstructor;
+    return getCanonicalDecl()->CXXConstructorDeclBits.IsInheritingConstructor;
   }
   size_t numTrailingObjects(OverloadToken<ExplicitSpecifier>) const {
-    return CXXConstructorDeclBits.HasTrailingExplicitSpecifier;
+    return getCanonicalDecl()
+        ->CXXConstructorDeclBits.HasTrailingExplicitSpecifier;
```
This change does not look quite correct.
Indeed `numTrailingObjects` is an internal detail of the object that should preserve its invariants. i.e, it is exactly whether an explicit specifier is **stored** as a trailing object of this object.

However, you are most probably on the right path.
In `ASTWriterDecl`, `addExplicitSpecifier(D->getExplicitSpecifier(), Record);` does indeed look incorrect. Because again, we are trying to serialize that object  - not parts that are stored in an other canonicl decl.

I think the solution might be as simple as replacing that line with `addExplicitSpecifier(D->getExplicitSpecifierInternal(), Record);` (`getExplicitSpecifierInternal()` does not call `getCanonicalDecl()`)

cor3ntin commented:
Thanks for working on this


This change needs a release note.
Please add an entry to `clang/docs/ReleaseNotes.rst` in the section the most adapted to the change, and referencing any Github issue this change fixes. Thanks!

koplas committed:
[PATCH] [clang][frontend] Fix serialization for CXXConstructorDecl (refs llvm#132794)

When retrieving the ExplicitSpecifier from a CXXConstructorDecl, one
of its canonical declarations is returned. To correctly write the
declaration record the ExplicitSpecifier of the current declaration
must be used.

Failing to do so results in a crash during deserialization.

cor3ntin reviewed (approved):
LGTM, Thanks! 
Will you need me to merge that for you?

koplas commented:
> LGTM, Thanks! Will you need me to merge that for you?

Yes, I don't have any permissions to merge this.

cor3ntin commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -316,6 +316,7 @@ Bug Fixes in This Version
 - Fixed a modules crash where exception specifications were not propagated properly (#GH121245, relanded in #GH129982)
 - Fixed a problematic case with recursive deserialization within ``FinishedDeserializing()`` where
   ``PassInterestingDeclsToConsumer()`` was called before the declarations were safe to be passed. (#GH129982)
+- Fixed a modules crash where an explicit Constructor was deserialized. (GH#132794)
```
```suggestion
- Fixed a modules crash where an explicit Constructor was deserialized. (#GH132794)
```

cor3ntin replied:
Can you also update the commit message to reflect the change better?

Commit:
Fix GitHub reference

Co-authored-by: cor3ntin <corentinjabot@gmail.com>

shafik reviewed (commented):
I am probably missing something here but I see you switched to using `getExplicitSpecifierInternal()` but in your summary you talk about using the correct AllocKind but I just don't see how they related at all.

I see the line before does `Record.push_back(D->getTrailingAllocKind());`. I looked at `ExplicitSpecifier` and also don't see how it relates.

Could you please clarify?

koplas commented:
@shafik, I updated the commit message based on the feedback of @cor3ntin. I have now updated the title and description to reflect the new commit message, to avoid further confusion.
